### PR TITLE
Remove IRazorFileProviderCache and rely on file expiration triggers to

### LINKED ISSUE #1969: Remove IRazorFileProviderCache and rely on file expiration triggers to file change expiry
The CompilerCache uses `IFileProvider.Watch` to know when files have changed in the filesystem. We can take advantage of this and avoid performing disk reads in the `VirtualPathRazorPageFactory` when the Razor page has not changed.

This is a follow up to #1708

### PR DESCRIPTION
**Pull request by pranavkm**
file change expiry.

Fixes #1969

### CODE REVIEW

**Comment by NTaylorMullen** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs [hunk 1]
```diff
@@ -156,30 +175,40 @@ private static bool Match(Type t)
...
         {
-            result = compile(file);
+            var compilationResult = compile(file);
 
-            var cacheEntry = new CompilerCacheEntry(file, result.CompiledType);
+            var cacheEntry = _cache.Set(normalizedPath,
+                                        new CompilerCacheEntry(file, compilationResult.CompiledType),
+                                        PopulateCacheSetContext);
 
             // Concurrent addition to MemoryCache with the same key result in safe race.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs]: nit: looks like this comment should be above the `cacheEntry` assignment

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCacheResult.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    /// <summary>
+    /// Result of <see cref="ICompilerCache"/>.
+    /// </summary>
+    public class CompilerCacheResult
+    {
+        /// <summary>
+        /// Result of <see cref="ICompilerCache"/> when the specified file does not exist in the
+        /// file system.
+        /// </summary>
+        public static readonly CompilerCacheResult FileNotFound = new CompilerCacheResult();
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCacheResult.cs]: make this a property instead

**Comment by dougbu** [comment]
:shipit: w/ field -> property change

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs]: :+1:

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCacheResult.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCacheResult.cs]: :+1:

**Pull request closed without merging**